Explain that hidden contents should not be referenced. (#3232)
modelica/ModelicaSpecification: chapters/annotations.tex

  Same as \lstinline!Access.packageText! and additionally the class, or part of the class, can be copied.
\end{enumerate}

The \lstinline!access! annotation holds for the respective class and all classes that are hierarchically on a lower level, unless overridden by a \lstinline!Protection! annotation with \lstinline!access!.
Overriding \lstinline!access=Access.hide! or \lstinline!access=Access.packageDuplicate! has no effect.

> Classes should not use other classes in ways that contradict this protection.
> For instance a class with \lstinline!Access.hide! should not be used in the diagram layer of a class with \lstinline!Access.diagram!, and there should not be hyperlinks to classes with \lstinline!Access.icon! (from classes with visible documentation).

Reviewer: There should be no room for interpretation:
```suggestion
Classes must not use other classes in ways that contradict this protection.
For instance, a class with \lstinline!Access.hide! must not be used in the diagram layer of a class with \lstinline!Access.diagram!, and there must not be hyperlinks to classes with \lstinline!Access.icon! from classes with visible documentation.
```

Reviewer: I wonder if we could go even further, and give the explicit rules that the generic formulation boils down to?
```suggestion
Classes must not use other classes in ways that contradict this protection:
\begin{itemize}
\item A class with \lstinline!Access.hide! or higher protection must not be used in the diagram layer of a class with \lstinline!Access.diagram!.
\item There must not be hyperlinks to classes with \lstinline!Access.icon! or higher protection from classes with visible documentation.
…
\end{itemize}
```

Author: I feared that the full list would be rather long, which would risk that something was missed and thus it was clearer to just give the goal.

The use of "should" instead of "must" was deliberate. If a package makes an error here that should not prevent us from simulating its models, so a strict requirement could be a problem.

Reviewer: As a library author defending for my intellectual property, I'd be disappointed if the specification didn't strictly require tools to obey access restrictions.  If they aren't set up correctly, the library should be fixed so that there is no ambiguity about how a tool should handle the contents.

Author: I agree that tools should obey the access restrictions, but the above is stated as restrictions on the classes - not on the tools.

So, to clarify add 
"Tools must ensure that protected contents is not shown, even if classes do not meet this requirement."

To me this "must" only makes sense if the restriction for the classes use "should", since "must" is an absolute requirement and here we are specifying what happens if the requirement is broken.

Note that this doesn't describe how the situation is handled by tools as long as the contents isn't shown - for diagrams Dymola have depending on circumstances used either Crossed-out-icons and just no indication; but the classes with those diagrams could still be used. (Obviously the specific libraries were corrected later.)

Reviewer: Sounds like a good discussion point for a web meeting.

Author: This was now added after the web-meeting.

  Same as \lstinline!Access.packageText! and additionally the class, or part of the class, can be copied.
\end{enumerate}

The \lstinline!access! annotation holds for the respective class and all classes that are hierarchically on a lower level, unless overridden by a \lstinline!Protection! annotation with \lstinline!access!.
Overriding \lstinline!access=Access.hide! or \lstinline!access=Access.packageDuplicate! has no effect.

Classes should not use other classes in ways that contradict this protection.

> Tools must ensure that protected contents is not shown, even if classes do not meet this requirement.

Reviewer 2: ```suggestion
Tools must ensure that protected content is not shown, even if classes do not meet this requirement.
```

Author: I'm not sure if that change is necessary, but I'm not totally against it.

"Contents" implies that it is more tangible and countable, and to me the individual elements are so tangible and countable and none of them should be shown.

However, I can understand "content" in the sense that it is the protected ideas represented there that shouldn't be shown.

Author: Yes. Done.

> For instance a class with \lstinline!Access.hide! should not be used in the diagram layer of a class with \lstinline!Access.diagram!, and there should not be hyperlinks to classes with \lstinline!Access.icon! (from classes with visible documentation).

Reviewer 2: Have the examples non-normative?

Author: Could possibly do, but there has also been a push for making it more normative.

Author: Added now.

  Same as \lstinline!Access.packageText! and additionally the class, or part of the class, can be copied.
\end{enumerate}

The \lstinline!access! annotation holds for the respective class and all classes that are hierarchically on a lower level, unless overridden by a \lstinline!Protection! annotation with \lstinline!access!.
Overriding \lstinline!access=Access.hide! or \lstinline!access=Access.packageDuplicate! has no effect.

> Classes should not use other classes in ways that contradict this protection.
> Tools must ensure that protected contents are not shown, even if classes do not meet this requirement.

Reviewer: Stronger:
```suggestion
A class may only be used in ways that comply with this protection.
Tools must ensure that protected contents are not revealed, even in case this rule about valid use is violated.
```

Author: The original text used should and must according to https://www.rfc-editor.org/rfc/rfc2119 - "may not" is not included in that list. (Well, "should" was possibly borderline.) Thus, I prefer to not change this. The other words could be changed.

Reviewer: It is this very meaning from RFC2119 that I find inappropriate in these statements:
> **[4]. SHOULD NOT**   This phrase, or the phrase "NOT RECOMMENDED" mean that there may exist valid reasons in particular circumstances when the particular behavior is acceptable or even useful, but the full implications should be understood and the case carefully weighed before implementing any behavior described with this label.

I don't think that the specification should open up for using the classes in ways that contradict the protection.  This way, the responsibility of the library developer is clear, but the second sentence still adds extra security for the library developer in case he/she is using a tool that fails to verify the protection requirements when preparing the library for distribution.

Author: > It is this very meaning from RFC2119 that I find inappropriate in these statements:
> 
> > **[4]. SHOULD NOT**   This phrase, or the phrase "NOT RECOMMENDED" mean that there may exist valid reasons in particular circumstances when the particular behavior is acceptable or even useful, but the full implications should be understood and the case carefully weighed before implementing any behavior described with this label.
> 
> I don't think that the specification should open up for using the classes in ways that contradict the protection. This way, the responsibility of the library developer is clear, but the second sentence still adds extra security for the library developer in case he/she is using a tool that fails to verify the protection requirements when preparing the library for distribution.

If it is an **absolute requirement**, I don't see that it makes sense to specify the behavior.
We could clarify that "The only acceptable use of class contradicting the protection is for testing the protection."

Reviewer: I agree that it would be best to not specify the behavior.  I would prefer to not clarify anything at all:
- The specification shouldn't say anything at all regarding what kind of testing tool vendors should perform in order to check their compliance with the specification.  On GitHub, we _could_ provide a test collection that can help tool vendors check their compliance, but it is not the specification's job to point out such resources.
- Instead of hinting that library authors perhaps should worry about how tools behave if the library violates the specification, we should explain the importance of library authors verifying that a tool meets their expectations when it comes to supporting correct `protection`-annotation behavior (for valid cases), before they allow their library to be decrypted by the tool.

Author: > I agree that it would be best to not specify the behavior.

I don't. The reason is that such issues occur does occur in practice.
However, if we don't want to specify the behavior it seems odd to make it an absolute requirement.

Author: One point is at _what level_ it shall be rejected.

If you encrypt library of 100 models, does a tool have to do some detailed(*) check that _all of them_ satisfy this before doing anything?

I would say not - and thus I see two realistic alternatives:
- Restrict what is shown for that model, but show other models
- Don't show the contents of that model, but show other models

Both are ok to me; but neither is consistent with viewing it as an absolute requirement.
Note that this may get complicated if the problematic model is used internally in the library.

*: There are also some non-detailed checks. E.g., if a tool cannot parse the library it doesn't make sense to present anything. However, as far as I understand SEMLA is designed so that even parsing of sub-parts can be delayed (similarly as for normal Modelica).

Author: This was resolved at the latest phone meeting.

\begin{nonnormative}

> For instance a class with \lstinline!Access.hide! should not be used in the diagram layer of a class with \lstinline!Access.diagram!, and there should not be hyperlinks to classes with \lstinline!Access.icon! (from classes with visible documentation).

Reviewer: Stronger:
```suggestion
For instance, a class with \lstinline!Access.hide! must not be used in the diagram layer of a class with \lstinline!Access.diagram!, and there must not be hyperlinks to classes with \lstinline!Access.icon! (from classes with visible documentation).
```

Author: Similarly.

Reviewer: Right, let's keep the discussion in the thread above.

> \end{nonnormative}

Reviewer: The rule about what tools should do when the usage rule has been violated would also deserve some illustration.
```suggestion
\end{nonnormative}

\begin{example}
Consider the following invalid use of a class with \lstinline!Access.hide!:
\begin{lstlisting}[language=modelica]
block MySecret
  RealOutput y = time;
  annotation(Protection(access = Access.hide));
end MySecret;

model M
  MySecret mySecret;
  Integrator integrator;
equation
  connect(mySecret.y, integrator.u);
  annotation(Protection(access = Access.diagram));
end M;
\end{lstlisting}
In order to not reveal the existence of the class \lstinline!MySecret!, a tool may choose to show the diagram with both \lstinline!mySecret! and all connections to it removed.
(The tool could also choose to not show the diagram of \lstinline!M! at all.)
When simulating \lstinline!M!, the tool must not store \lstinline!mySecret.y!

Had \lstinline!M! had \lstinline!Access.packageDuplicate! instead, one way for a tool to not reveal the existence of \lstinline!MySecret! would be to refuse to show the class definition of \lstinline!M! at all.
\end{example}
```

Author: I believe it is important to show that M and MySecret are part of the same encrypted package, so enclose them in package P (no need to have a longer name). For the final sentence I would just completely remove it or reformulate it based on the other example:

If a user entered
```
model My
  P.MySecret a;
end My;
```
That should fail similarly as if `P.MySecret` did not exist, and especially the icon for `P.MySecret` should not be used.

-
Basically a class may be revealed in Modelica text, in documentation, in comments, or just guessed. As soon as we say that a hidden class is different from a non-existent class we open up a way to reveal their existence.

Thus trying to force tools to treat hidden classes in badly designed package specially we would force tools to reveal the existence of hidden classes in normal cases; that doesn't make sense.

Reviewer: > That should fail similarly as if P.MySecret did not exist, and especially the icon for P.MySecret should not be used.

This makes sense.  I'll use it to reformulate the example.

Overriding \lstinline!access=Access.hide! or \lstinline!access=Access.packageDuplicate! has no effect.

\begin{example}
If the annotation is given on the top level of a package and at no other class in this package, then the \lstinline!access! annotation holds for all classes in this package.
\end{example}

Classes should not use other classes in ways that contradict this protection.
Tools must ensure that protected contents are not shown, even if classes do not meet this requirement.

\begin{example}
For instance a class with \lstinline!Access.hide! should not be used in the diagram layer of a class with \lstinline!Access.diagram!, and there should not be hyperlinks to classes with \lstinline!Access.icon! (from classes with visible documentation).

Consider the following invalid use of a class with \lstinline!Access.hide!:
\begin{lstlisting}[language=modelica]
package P
  block MySecret
    RealOutput y = time;
    annotation(Protection(access = Access.hide));
  end MySecret;
  model M

>     MySecret mySecret;
>     Integrator integrator;
>   equation
>     connect(mySecret.y, integrator.u);

Reviewer: To fully illustrate the point of the example, one should also try to place the icons on the diagram:
```suggestion
    MySecret mySecret
     annotation(Placement(
       transformation(origin = {30, 30}, extent = {{-10, -10}, {10, 10}})));
    Integrator integrator
     annotation(Placement(
       transformation(origin = {70, 30}, extent = {{-10, -10}, {10, 10}})));
  equation
    connect(mySecret.y, integrator.u)
      annotation(Line(origin = {49.5, 30}, points = {{-8.5, 0}, {8.5, -0}}));
```

Author: To me this use of origin and extent for seems like a weird style, but as stated before the specification shouldn't try to enforce a style-guide.

Author: > For what it's worth, I think System Modeler does it this way as a way of giving the user control of the center of rotation used in GUI rotation operations. 

I believe Dymola introduces the origin (originally in the center) when you start rotating.

Reviewer: Right, that should work too.

    annotation(Protection(access = Access.diagram));
  end M;

> end P;

Reviewer: The problematic case of even less protection within the package should also be discussed:
```suggestion
  model M2
    MySecret mySecret;
    annotation(Protection(access = Access.packageDuplicate));
  end M2;
end P;
```

Reviewer: There are too many open suggestions right now for also suggesting the text to go with it:
> Despite the protection level of \lstinline!P.M2!, a too must be cautious to not reveal the text with enough detail to tell that there is a reference to the class \lstinline!P.MySecret! (this is different compared to when such misuse of a protected class occurs outside of the package containing the protected class).

Author: I understand the concern, but one could also just have a text saying: "There is a hidden class MySecret" in the documentation, or having `//MySecret mySecret;` in the code.

(Technically using `MySecret` in an exposed class inside the protected package doesn't necessarily imply that there's a class called `MySecret`, as the package might have `import MySecret=Modelica.Blocks.Math.Gain;`)

\end{lstlisting}

> In order to not reveal the existence of the class \lstinline!P.MySecret!, a tool may choose to show the diagram with both \lstinline!mySecret! and all connections to it removed.

Reviewer: Clarify which diagram:
```suggestion
In order to not reveal the existence of the class \lstinline!P.MySecret!, a tool may choose to show the diagram of \lstinline!P.M! with both \lstinline!mySecret! and all connections to it removed.
```

> (The tool could also choose to not show the diagram of \lstinline!P.M! at all.)

Reviewer: Possibly to be combined with the sentence on the next list:
```suggestion
As long as the invalid use of \lstinline!P.MySecret! occurs within the same top level package as where the class is defined (here, \lstinline!P!), a tool is allowed to silently ignore the use for purposes of model translation.
```

> When simulating \lstinline!P.M!, the tool must not store \lstinline!mySecret.y!

Reviewer: Punctuation:
```suggestion
When simulating \lstinline!P.M!, the tool must not store \lstinline!mySecret.y!.
```

> 
> If a user enters
> \begin{lstlisting}[language=modelica]
> model My
>   P.MySecret a;
> end My;
> \end{lstlisting}
> That icon for \lstinline!P.MySecret! shall not be shown, similarly as if the class did not exist.

Reviewer: Make this a separate example with focus on lookup behavior:
```suggestion
In order to support development of valid protected packages, it is of course OK and expected that a tool will report the invalid use of \lstinline!P.MySecret! in \lstinline!P.M! (revealing its existence in a diagnostic) during development of the package.
\end{example}

\begin{example}
With the same package \lstinline!P! as in the previous example, consider invalid outside of \lstinline!P!:
\begin{lstlisting}[language=modelica]
model My
  P.MySecret a
    annotation(Placement(
       transformation(origin = {30, 30}, extent = {{-10, -10}, {10, 10}})));;
end My;
\end{lstlisting}
Regardless of the protection of \lstinline!My!, a tool must act as if \lstinline!P.MySecret! did not exist.
For example, translation of \lstinline!My! must fail with a standard error message about reference to the non-existing class \lstinline!P.MySecret!.
Further, just like when being misused inside the package \lstinline!P!, the tool must not reveal that it knows about the icon of \lstinline!P.MySecret!.
With such precautions taken, showing the text or diagram of \lstinline!My! is permitted as it doesn't reveal the actual existence of \lstinline!P.MySecret!.
```
